src/SFML/Graphics/RenderTextureImplFBO.cpp

```diff
@@ -513,17 +515,28 @@ bool RenderTextureImplFBO::activate(bool active)
 
     Uint64 contextId = Context::getActiveContextId();
 
+    // FIXME: This does not work because sometimes Context::getActiveContext()
+    // returns a different context than Context::getActiveContextId()
+    const Context* activeContext = Context::getActiveContext();
+    if (m_sRgb && (!activeContext || !activeContext->getSettings().sRgbCapable))
+    {
+        contextId = 0;
+    }
```

Comment:
@binary1248 Here is some code where I need to know the effective settings of the active context in order to tell if I need to switch to a sRGB compatible one before rendering. I think I am interested in the internal active context that "manage unshareable resources" not the "outside facing" context.
However since `Context::getActiveContext()` is not the same as `Context::getActiveContextId()`, my code does not work. On my machine, even when it would be required for correct output, it makes the texture never use its own sRGB capable rendering context except during startup.
@eXpl0it3r told me that you are the maintainer aware of these tricky details, and I need some advice on how to finish this properly.

Reply:
In my last commit, I provided a fix with a new method `Context::getActiveContextSettings()`. There might be other ways, but none seemed more satisfying.

Reply:
I don't understand why the context has to actually have an sRGB pixel format surface (the surface format is what is reported by the context settings after creation). The context used or activated by the `RenderTextureImplFBO` is merely a container to house the FBO since it is an unshareable resource. The surface of this context will essentially never be written to (we write to the texture attachment). When `GL_FRAMEBUFFER_SRGB` is enabled the conversion should still take place independent of the pixel format of a context. Isn't it possible to just enable `GL_FRAMEBUFFER_SRGB` for whatever context gets activated instead of jumping through hoops to get an sRGB context surface that we will never write to?

Reply:
As you said we do not care about the surface of the context. However we cannot mess up with the surface of the currently activated context. Maybe it is only a Linux problem but on my machine, if I enable `GL_FRAMEBUFFER_SRGB` on a context, it will consider that its surface must be sRGB encoded, which gives unexpected results.
It seemed to me that there is no easy ways of enabling `GL_FRAMEBUFFER_SRGB` just for the `RenderTextureImplFBO` since there is no deactivation hook, and the alternative would be to call `glEnable`/`glDisable` on every draw.

Just to be clear, this situation where we must create the useless surface is niche anyway. I expect very few users to actually request a sRGB texture with a non sRGB surface. Although it is not a reason for it to be buggy, if there can be a less optimized case, it is this one.

Reply:
From https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_framebuffer_sRGB.txt

> If FRAMEBUFFER_SRGB_EXT is enabled __and the boolean
    FRAMEBUFFER_SRGB_CAPABLE_EXT state for the drawable is true__, the R,
    G, and B destination color values (after conversion from fixed-point
    to floating-point) are considered to be encoded for the sRGB color
    space and hence need to be linearized prior to their use in blending.
...
Otherwise if FRAMEBUFFER_SRGB_EXT is disabled, __or the
    drawable is not sRGB capable__, or the value corresponds to the A
    component, then cs = cl for such components.

This wording was adapted into the core specification at some point as well.

In theory one could leave `GL_FRAMEBUFFER_SRGB` enabled for all contexts in the whole application and OpenGL should intelligently recognize when it makes sense to perform the conversion (when the drawable is sRGB capable). If you enable `GL_FRAMEBUFFER_SRGB` and try to draw to a non sRGB surface/drawable it will/should have the same effect as if `GL_FRAMEBUFFER_SRGB` was not enabled.

Therefore I don't see the necessity to mess with or worry about how a context was created. What matters is the FBO attachment (has to be sRGB) and whether `GL_FRAMEBUFFER_SRGB` is enabled at the time of drawing. Nothing else should matter, assuming the implementation conforms to the specification of course.

Since `GL_FRAMEBUFFER_SRGB` should be a simple flag somewhere deep in the driver, I also wouldn't be concerned about having to enable/disable it whenever drawing to the FBO. Sure... I myself keep saying every call to the driver is one call too many, but as it stands there are many other worse offenders in the codebase.

Reply:
Problem is, in practice, even when not requesting sRGB capable surface, I get one. Is there an easy way to disable that ? I agree it would be better. The spec allows us not to worry about disabling `GL_FRAMEBUFFER_SRGB` for non-sRGB render texture though.

About enable/disable, the problem is that I have no easy way to tell that we are finishing an FBO rendering that explicitly activated `GL_FRAMEBUFFER_SRGB`. If that is what you want, maybe you have a clue on how to do that.

Reply:
https://github.com/SFML/SFML/blob/master/src/SFML/Graphics/Texture.cpp#L204

The texture image is only created with sRGB storage if it is explicitly requested. Not requesting an sRGB `RenderTexture` should lead to not requesting an sRGB texture which should lead to no sRGB conversion taking place.

If it is true that you are getting an sRGB texture even when one really shouldn't be created then the reason behind this should be investigated.

Remember... only the target drawable is relevant. The surface format of the context itself is (or should be) completely irrelevant in this context according to the specification since we are not targeting it with our draw.

As for enable/disable, it might seem "messy" but the way SFML deals with the issue is it just makes sure the required state is set right before drawing. In this case you just leave it enabled and the next place where a draw takes place you disable it (if the target is sRGB and you don't want a conversion to happen for some reason). If it turns out that this leads to severe performance degradation, the enabled/disabled state can be tracked in `RenderTarget` as is already done for some other things. Application-wide OpenGL state tracking is something to be looked into in a future re-implementation of the graphics module.

Reply:
Sorry I was not clear, for RenderTextures I get the requested sRGB storage and this works perfectly well.
However for window surfaces, I always get an sRGB surface even when I do not ask for it.

Reply:
Can you post a list of visuals that are provided by your system via the `glxinfo` command? Maybe we might need to tweak the visual selection to favour a non-sRGB visual if the user doesn't request one.

Reply:
Here is the output of `glxinfo` on my machine: https://gist.github.com/Sakarah/38707bbd68d82e68286ae0424d66aca6
It seems that I have both sRGB and non-sRGB visuals available.

Reply:
You can try adding a cost to the inverse case at https://github.com/SFML/SFML/blob/master/src/SFML/Window/GlContext.cpp#L656

When the user requests no sRGB and the format is sRGB it should also be disfavoured.

Reply:
With a cost for the inverse case SFML selects visual `0x582` with sRGB and visual `0x50a` without. However, if I enable `GL_FRAMEBUFFER_SRGB`, even with the visual `0x50a` (which should be non-sRGB) it will sRGB encode. It might be a bug in my graphics driver since unless I misunderstood OpenGL spec, it should not.

Reply:
According to many articles, sRGB support was really buggy in the past (e.g. https://www.g-truc.net/post-0720.html). I can't tell if it has gotten any better nowadays, however I wouldn't be surprised if it didn't since OpenGL has unintentionally become somewhat of a legacy API now that Khronos is spending more of their time on Vulkan.

If this is really still the case, we would have to implement this "intelligence" that should be in the driver in our own code. That can get somewhat complicated... and dirty...

Reply:
My driver clearly seem to follow the rule stated in the article you quote:
> IF FRAMEBUFFER_BINDING is 0 then if FRAMEBUFFER_SRGB is enabled, the R, G, and B values after blending are converted into the non-linear sRGB color space. Otherwise, follow the specification.

I guess it means we need to use the enable/disable strategy then.

Reply:
Just a thought... 

Would it make sense to add another virtual member function to `RenderTarget` so that it can know if the concrete implementation has an sRGB target or not? Something like `virtual bool RenderTarget::isSrgb() const`. This would have to be overridden by `RenderWindow` and `RenderTexture`. `RenderTexture` would have to decide based on 1) the concrete Impl and 2) the target of the Impl.

The `RenderTarget` would use this information to enable/disable `GL_FRAMEBUFFER_SRGB` alongside all the other state management it already does now.

Reply:
Might be interesting but can we do that without introducing a breaking change ?
I don't think there is a good default for `virtual bool RenderTarget::isSrgb() const` that do not force users to adapt their code if they have a custom child class for RenderTarget.

Reply:
Can't we just assume the default is to return `false`? The function doesn't have to be pure virtual. That way it wouldn't break existing code. In the off chance users have implemented their own `RenderTarget` and it is even sRGB, they will be happy to have a mechanism to inform the draw code of this information.

Reply:
It seems to me that both defaulting to `true` and to `false` will make a custom `RenderTarget` behave differently. If `true` then, we insert a call to `glEnable(GL_FRAMEBUFFER_SRGB)` before a draw batch, if false, we insert a call to `glDisable(GL_FRAMEBUFFER_SRGB)`. Depending on user code, both may break the drawing.
Therefore I currently think that a pure virtual is still a better choice, to make the breaking change generate an error instead of silently failing to correctly draw.
If SFML has a strict breaking change policy, we might prefer to find another workaround, such as using the `setActive` overrides to change the `GL_FRAMEBUFFER_SRGB` flag.

Reply:
Generally, it has always been the case that you can manipulate OpenGL state within any sfml-graphics objects "at your own risk". SFML only guarantees that everything will work as long as all your OpenGL stuff is not done in the same context that is used by any sfml-graphics objects. As such, as weird as it sounds, behaviour related to enabling/disabling `GL_FRAMEBUFFER_SRGB` in user code was never guaranteed and thus cannot be broken by definition.

Reply:
I just pushed a new version using this last idea